dkmgt-fwutil: New tool to manipulate TP-Link Omada firmware files

oskirby opened the pull request:
This adds a new tool to create TP-Link Omada firmware files. This was created mostly by reverse engineering the firmware and bootloader on an ER8411. It also seems to be compatible with the firmware for the ER605 and possibly other models too.

Although the firmware contains an RSA signature, the bootloader's recovery mode doesn't seem to verify it, meaning that this can be used to create factory installation images through the recovery bootloader.

It's probably worth calling out that this adds the [cJSON](https://github.com/DaveGamble/cJSON) library to the project, which was used to parse/manipulate JSON partitions embedded in the firmware file. This library is licensed under the MIT license, which I believe is compatible with this project. If that's not the case, I can probably rewrite the JSON handling to use some other library that is more appropriate.

oskirby committed:
dkmgt-fwutil: New tool to manipulate TP-Link Omada firmware files

This adds a new tool to create and manipulate TP-Link Omada firmware
files. This was created mostly by reverse engineering the firmware
and bootloader on an the TP-Link ER8411. It also seems to be
compatible with the firmware for the ER605 and possibly other models
too.

Although the firmware contains an RSA signature, the bootloader doesn't
seem to verify it, meaning that this can be used to create factory
installation images via the bootloader's recovery mode.

Signed-off-by: Naomi Kirby <dev@manawolf.ca>

dangowrt commented:
Imho it would be better to add cJSON as tools library (ie. a dedicated package building the shared library, similar to eg. `tools/libdeflate`). This would prevent us from duplicating the code, and make maintenance easier in the long run (eg. automated tools we are using would point out the library being outdated, and all we'd have to do in this case would be to edit the version and hash in `tools/cjson/Makefile`).
Other than that the dkmgt tool itself looks fine in a quick scroll-over. Maybe the commenting style could be improved a bit, ie. not need for those whole lines of `**********`, and I'd say either use `//` comments or use `/* */` style. But many of the tools in this repo look much worse, so I wouldn't hesitate to just merge it once cJSON is properly (tools-) packaged instead of embedding a copy of it.

oskirby committed:
Link to cJSON library instead of vendoring a copy.

Signed-off-by: Naomi Kirby <dev@manawolf.ca>

oskirby committed:
Update comment style to match Linux kernel

Some PR feedback was to try and match the Linux kernel commenting style
which means we should eliminate the full-line '*****' separators and
the use of C++ style comments.

Signed-off-by: Naomi Kirby <dev@manawolf.ca>

oskirby committed:
Refine the default action handling

By moving the default action handling to the end of the program, we
can ensure that an explicitly specified action (eg: '--print') always
takes precedence over an implicit one (eg: writing the firmware after
an update to some partition).

Signed-off-by: Naomi Kirby <dev@manawolf.ca>

oskirby committed:
Add dkmgt_ptn_parse_json helper function

This eliminates some duplicated code in handling JSON content embedded
in partitions, and suppresses some bogus error messages when extracting
firmware files.

Signed-off-by: Naomi Kirby <dev@manawolf.ca>

kuanyili commented on src/dkmgt-fwutil.h:
```diff
@@ -0,0 +1,83 @@
+#define DKMGT_MAGIC_0       0xa5a5a5a5
+#define DKMGT_MAGIC_1       0x8f72632a
+#define DKMGT_MAGIC_2       0x40f8600
+#define DKMGT_MAGIC_3       0x9206b51
```
Shall be `0x9206bb51`? Maybe add a lead zero to the previous line (`0x040f8600`) so missing digits would be obvious.
